languages/rust/exercises/concept/tuples/.docs/instructions.md

```diff
@@ -0,0 +1,48 @@
+# Instructions
+
+You are working on a game targeting a low-power embedded system and need to write several convenience functions which will be used by other parts of the game.
+
+## 1. Write a function `divmod` which returns both the quotient and remainder of a division
+
+```rust
+fn divmod(dividend: i16, divisor: i16) -> (i16, i16)
+```
+
+Example:
+
+```rust
+assert_eq!(divmod(10, 3), (3, 1));
+```
+
+## 2. Write an iterator adaptor `evens` which returns the even items from an arbitrary iterator
+
+This will be helpful to enable a screen-buffer optimization, your boss promises.
+
+[Iterators](https://doc.rust-lang.org/std/iter/trait.Iterator.html) are a separate concept, and we aren't going to get too in-depth about their syntax or mechanics here. The important facts about them for this task:
```

Comment:
Unfortunately, this is not how v3 exercise are to be designed. One can’t say “Iterators are a separate concept, and we aren’t going to get too in-depth about their syntax or mechanics here.” and then ask someone to use the. The alternatives are either to teach them as a concept in this exercise, remove them, or have them as a prerequisite.

Reply:
This is a hard one to work around. This is the third attempt at implementing a tuples exercise, because we keep running into this issue: tuples are fundamental, and should be one of the first concepts a student encounters. However, there is basically no plausible reason to use them which doesn't involve _something_ which comes later in the concept graph.

This exercise doesn't teach iterators at all. Removing them means entirely rewriting the portion of the exercise teaching destructuring binding. Making them prerequisite is in some way putting the cart before the horse; we're very likely to end up with dependency cycles that way. 

This exercise is written such that the student doesn't need to know anything about iterators except that they support certain methods, all of which are mentioned and have linked-to documentation. Is that really too much?

Reply:
Perhaps you could not bring up that there's more to iterators at all. Something along the lines of

```md
Iterators are constructs [I don't know the correct term in Rust lingo] that meet the following requirements:
- An iterator is an arbitrary-length stream of items
- They have an `enumerate` method, ...
- ...
```

A definition of the iterator concept that is not incorrect but also not complete. Remove the "don't worry too much about it", just leave out everything that's not strictly needed here.

I've seen this technique used a lot in maths or physics lectures, where often you need a definition to work with, but then (much) later it turns out that this is really just a special case of a more general concept, or it's not a definition itself but can be derived from other definitions instead.

That way you might work around the issue @iHiD outlined in the other comment because people _won't know_ there's more to it, so they don't feel compelled to get distracted. They may still suspect that there's more to them, but at some point you just need to consider that as their own problem (similar to how you won't explain how the compiler works in a hello world example, and how the result is translated to currents on a piece of rock, even if that's technically necessary to get a complete understanding of the example).

Reply:
Yeah. This is a really good suggestion, @SaschaMann. We would then have this exercise teach the concept of `iterators`, but only scrape the surface. Future exercises might then introduce the more complex additions to iterations as separate concepts `iterators-xxx`.

I will think more on this throughout the day though.

Reply:
What do you think of the description in 3f1416b8? It explains the syntax on a surface level without really getting into the notion of what trait objects are, let alone generics.

I really don't think this exercise should introduce the `iterators` concept, because it really doesn't teach them; they're present but unexplained. There's plenty of room for a separate exercise focusing on them.

Reply:
I think it works (other than the minor comment I left). It avoids creating a feeling of "there's more to this" imo.